File: ios/Jazzify/EarTraining/Engine/EarTrainingPrecisionMidi.swift
import Foundation

/// Web `earTrainingPrecisionMidi.ts` 相当。
enum EarTrainingPrecisionMidi {
    private struct ParsedMidiNoteEvent {
        let startTick: Int
        let endTick: Int
        let midi: Int
    }

    private struct ActiveNote {
        let startTick: Int
    }

    static func buildFromMidi(
        data: Data,
        bpm: Int,
        transposeOffset: Int = 0
    ) throws -> EarTrainingPrecisionNoteBuildResult {
        let tempoMap = try EarTrainingSmfBinary.buildTempoMap(data)
        let (_, trackChunks) = try EarTrainingSmfBinary.parseHeader(data)
        var parsed: [ParsedMidiNoteEvent] = []
        for chunk in trackChunks {
            parsed.append(contentsOf: parseTrackNoteEvents(data: data, trackStart: chunk.trackStart, trackEnd: chunk.trackEnd))
        }

        var clusterIndexByTick: [Int: Int] = [:]
        var notes: [EarTrainingPrecisionNote] = []
        notes.reserveCapacity(parsed.count)

        for event in parsed {
            let startSec = EarTrainingSmfBinary.tickToSeconds(tempoMap, tick: event.startTick)
            let endSec = EarTrainingSmfBinary.tickToSeconds(tempoMap, tick: event.endTick)
            let durationSec = max(0.05, endSec - startSec)
            let midi = event.midi + transposeOffset
            let indexInCluster = clusterIndexByTick[event.startTick] ?? 0
            clusterIndexByTick[event.startTick] = indexInCluster + 1
            notes.append(
                EarTrainingPrecisionNote(
                    id: "m:\(event.startTick):\(midi):\(indexInCluster)",
                    midi: midi,
                    startSec: startSec,
                    durationSec: durationSec,
                    isBlackKey: EarTrainingPrecisionNotes.isBlackKeyMidi(midi),
                    measureNumber: 1,
                    isShortNote: EarTrainingPrecisionNotes.isShortNoteDuration(durationSec: durationSec, bpm: bpm)
                )
            )
        }

        notes.sort { lhs, rhs in
            if abs(lhs.startSec - rhs.startSec) > 0.0005 {
                return lhs.startSec < rhs.startSec
            }
            if lhs.midi != rhs.midi {
                return lhs.midi < rhs.midi
            }
            return lhs.id < rhs.id
        }

        let trimmedNotes = EarTrainingPrecisionNotes.trimOverlappingSamePitchNotes(
            notes: notes,
            classificationBpm: bpm
        )

        let keyboardRange = EarTrainingPrecisionNotes.resolveKeyboardRange(noteMidis: trimmedNotes.map(\.midi))
        return EarTrainingPrecisionNoteBuildResult(notes: trimmedNotes, keyboardRange: keyboardRange)
    }

    private static func parseTrackNoteEvents(
        data: Data,
        trackStart: Int,
        trackEnd: Int
    ) -> [ParsedMidiNoteEvent] {
        var notes: [ParsedMidiNoteEvent] = []
        var activeByKey: [String: [ActiveNote]] = [:]
        var offset = trackStart
        var absoluteTick = 0
        var runningStatus = 0

        func closeNote(channel: Int, midi: Int, endTick: Int) {
            let key = "\(channel):\(midi)"
            guard var stack = activeByKey[key], !stack.isEmpty else { return }
            let started = stack.removeLast()
            if stack.isEmpty {
                activeByKey.removeValue(forKey: key)
            } else {
                activeByKey[key] = stack
            }
            if endTick > started.startTick {
                notes.append(ParsedMidiNoteEvent(startTick: started.startTick, endTick: endTick, midi: midi))
            }
        }

        func pushNoteOn(channel: Int, midi: Int, velocity: Int) {
            if velocity <= 0 {
                closeNote(channel: channel, midi: midi, endTick: absoluteTick)
                return
            }
            let key = "\(channel):\(midi)"
            var stack = activeByKey[key] ?? []
            stack.append(ActiveNote(startTick: absoluteTick))
            activeByKey[key] = stack
        }

        while offset < trackEnd {
            let delta = EarTrainingSmfBinary.readVariableLengthQuantity(data, offset: offset)
            absoluteTick += delta.value
            offset = delta.nextOffset
            guard offset < trackEnd else { break }

            var status = Int(data[offset])
            if status < 0x80 {
                status = runningStatus
            } else {
                offset += 1
            }

            if status == 0xff {
                guard offset < trackEnd else { break }
                let metaType = Int(data[offset])
                offset += 1
                let len = EarTrainingSmfBinary.readVariableLengthQuantity(data, offset: offset)
                offset = len.nextOffset + len.value
                if metaType == 0x2f {
                    break
                }
                continue
            }

            if status == 0xf0 || status == 0xf7 {
                let len = EarTrainingSmfBinary.readVariableLengthQuantity(data, offset: offset)
                offset = len.nextOffset + len.value
                continue
            }

            let high = status & 0xf0
            let channel = status & 0x0f
            runningStatus = status

            if high == 0x90 {
                guard offset + 1 < trackEnd else { break }
                let midi = Int(data[offset])
                let velocity = Int(data[offset + 1])
                pushNoteOn(channel: channel, midi: midi, velocity: velocity)
                offset += 2
                continue
            }

            if high == 0x80 {
                guard offset < trackEnd else { break }
                let midi = Int(data[offset])
                closeNote(channel: channel, midi: midi, endTick: absoluteTick)
                offset += 2
                continue
            }

            offset += EarTrainingSmfBinary.dataBytesForChannelStatus(status)
        }

        return notes
    }
}

File: src/utils/earTrainingPrecisionMidi.ts
import {
  isPrecisionShortNoteDuration,
  resolvePrecisionKeyboardRange,
  trimOverlappingSamePitchNotes,
  type PrecisionNote,
  type PrecisionNoteBuildResult,
} from '@/utils/earTrainingPrecisionNotes';
import {
  buildSmfTempoMap,
  parseSmfHeader,
  readSmfVariableLengthQuantity,
  smfDataBytesForChannelStatus,
  smfTickToSeconds,
} from '@/utils/smfBinary';

const BLACK_KEY_OFFSETS = new Set([1, 3, 6, 8, 10]);

const isBlackKeyMidi = (midi: number): boolean => (
  BLACK_KEY_OFFSETS.has(((Math.round(midi) % 12) + 12) % 12)
);

const precisionMidiNoteId = (startTick: number, midi: number, indexInCluster: number): string => (
  `m:${startTick}:${midi}:${indexInCluster}`
);

interface ParsedMidiNoteEvent {
  startTick: number;
  endTick: number;
  midi: number;
}

interface ActiveNote {
  startTick: number;
}

const parseTrackNoteEvents = (
  data: Uint8Array,
  trackStart: number,
  trackEnd: number,
): ParsedMidiNoteEvent[] => {
  const notes: ParsedMidiNoteEvent[] = [];
  const activeByKey = new Map<string, ActiveNote[]>();
  let offset = trackStart;
  let absoluteTick = 0;
  let runningStatus = 0;

  const pushNoteOn = (channel: number, midi: number, velocity: number): void => {
    if (velocity <= 0) {
      closeNote(channel, midi, absoluteTick);
      return;
    }
    const key = `${channel}:${midi}`;
    const stack = activeByKey.get(key) ?? [];
    stack.push({ startTick: absoluteTick });
    activeByKey.set(key, stack);
  };

  const closeNote = (channel: number, midi: number, endTick: number): void => {
    const key = `${channel}:${midi}`;
    const stack = activeByKey.get(key);
    if (!stack || stack.length === 0) {
      return;
    }
    const started = stack.pop();
    if (!started) {
      return;
    }
    if (stack.length === 0) {
      activeByKey.delete(key);
    }
    if (endTick > started.startTick) {
      notes.push({
        startTick: started.startTick,
        endTick,
        midi,
      });
    }
  };

  while (offset < trackEnd) {
    const delta = readSmfVariableLengthQuantity(data, offset);
    absoluteTick += delta.value;
    offset = delta.nextOffset;

    let status = data[offset];
    if (status < 0x80) {
      status = runningStatus;
    } else {
      offset += 1;
    }

    if (status === 0xff) {
      const metaType = data[offset];
      offset += 1;
      const len = readSmfVariableLengthQuantity(data, offset);
      offset = len.nextOffset + len.value;
      if (metaType === 0x2f) {
        break;
      }
      continue;
    }

    if (status === 0xf0 || status === 0xf7) {
      const len = readSmfVariableLengthQuantity(data, offset);
      offset = len.nextOffset + len.value;
      continue;
    }

    const high = status & 0xf0;
    const channel = status & 0x0f;
    runningStatus = status;

    if (high === 0x90) {
      const midi = data[offset];
      const velocity = data[offset + 1];
      pushNoteOn(channel, midi, velocity);
      offset += 2;
      continue;
    }

    if (high === 0x80) {
      const midi = data[offset];
      closeNote(channel, midi, absoluteTick);
      offset += 2;
      continue;
    }

    offset += smfDataBytesForChannelStatus(status);
  }

  return notes;
};

export const buildPrecisionNotesFromMidi = (
  data: Uint8Array,
  bpm: number,
  transposeOffset = 0,
): PrecisionNoteBuildResult => {
  const tempoMap = buildSmfTempoMap(data);
  const { trackChunks } = parseSmfHeader(data);
  const parsed: ParsedMidiNoteEvent[] = [];
  for (const chunk of trackChunks) {
    parsed.push(...parseTrackNoteEvents(data, chunk.trackStart, chunk.trackEnd));
  }

  const clusterIndexByTick = new Map<number, number>();
  const notes: PrecisionNote[] = parsed.map(event => {
    const startSec = smfTickToSeconds(tempoMap, event.startTick);
    const endSec = smfTickToSeconds(tempoMap, event.endTick);
    const durationSec = Math.max(0.05, endSec - startSec);
    const midi = event.midi + transposeOffset;
    const indexInCluster = clusterIndexByTick.get(event.startTick) ?? 0;
    clusterIndexByTick.set(event.startTick, indexInCluster + 1);
    return {
      id: precisionMidiNoteId(event.startTick, midi, indexInCluster),
      midi,
      startSec,
      durationSec,
      isBlackKey: isBlackKeyMidi(midi),
      measureNumber: 1,
      isShortNote: isPrecisionShortNoteDuration(durationSec, bpm),
    };
  });

  notes.sort((a, b) => {
    if (Math.abs(a.startSec - b.startSec) > 0.0005) {
      return a.startSec - b.startSec;
    }
    if (a.midi !== b.midi) {
      return a.midi - b.midi;
    }
    return a.id.localeCompare(b.id);
  });

  const trimmedNotes = trimOverlappingSamePitchNotes(notes, bpm);

  const keyboardRange = resolvePrecisionKeyboardRange(trimmedNotes.map(note => note.midi));
  return { notes: trimmedNotes, keyboardRange };
};
